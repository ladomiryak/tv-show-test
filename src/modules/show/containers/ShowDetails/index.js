import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import Component from "./ShowDetails";
import * as showActions from "../../actions";
import { showSelectors } from "../../selectors";

const mapStateToProps = (state) => ({
  show: showSelectors.getShow(state),
  error: showSelectors.getError(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      fetchShow: showActions.fetchShow,
    },
    dispatch
  );

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(Component);

export default ConnectedComponent;
