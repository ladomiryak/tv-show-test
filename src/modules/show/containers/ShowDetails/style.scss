@import "../../.././../assets/scss/colors";

.show-details {
  margin-bottom: 40px;
  border-bottom: 1px solid #eee;
  padding: 40px 0;

  @media screen and (min-width: 768px) {
    display: flex;
  }

  &__poster {
    margin-right: 32px;
    border-radius: 4px;
    box-shadow: 0 0 10px 5px rgba($primary, 0.1);
    min-width: 300px;
    max-height: 500px;
    object-fit: cover;
    margin-bottom: 40px;
  }

  &__title {
    margin-bottom: 24px;
    font-size: 24px;
    font-weight: bold;
    color: $primary;
  }

  &__genres {
    display: inline-flex;
    margin-bottom: 8px;
    color: $tertiary;

    li {
      margin-right: 4px;
    }
  }

  &__description {
    line-height: 1.5;
    max-width: 95%;
    color: $primary;
  }
}
