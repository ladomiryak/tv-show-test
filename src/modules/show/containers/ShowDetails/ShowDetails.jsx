import React, { useEffect } from "react";
import { Loader, Alert } from "../../../shared";
import "./style.scss";

const ShowDetails = ({ error, showId, show, fetchShow }) => {
  // fetch Show Details
  useEffect(() => {
    if (showId && !show) {
      fetchShow({
        showId,
      });
    }
  }, [showId, show, fetchShow]);

  if (error) {
    return (
      <Alert variant="error">Something went wrong. Try later, please.</Alert>
    );
  }

  if (!show) {
    return <Loader />;
  }

  const {
    name,
    summary,
    image: { original },
    rating: { average },
    premiered,
    genres,
  } = show;

  return (
    <section className="show-details">
      <img className="show-details__poster" src={original} alt="Show poster" />
      <div>
        <h1 data-testid="show-name" className="show-details__title">
          {name} ({premiered.split("-")[0]}) {average}
        </h1>
        <ul className="show-details__genres">
          {genres.map((genre, i) => (
            <li key={genre}>
              {genre}
              {i < genres.length - 1 ? "," : ""}
            </li>
          ))}
        </ul>
        <p
          className="show-details__description"
          dangerouslySetInnerHTML={{ __html: summary }}
        />
      </div>
    </section>
  );
};

export default ShowDetails;
