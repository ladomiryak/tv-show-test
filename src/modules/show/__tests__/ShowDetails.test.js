import React from "react";
import ShowDetails from "../containers/ShowDetails";
import { render, screen } from "../test-util";
import stateMock from "./mock.json";

describe("Show Details", () => {
  test("Loader should be shown with empty state", () => {
    render(<ShowDetails />, {
      initialState: { show: {} },
    });

    expect(screen.getByTestId("loader")).toBeInTheDocument();
  });

  test("Title should be shown", () => {
    const state = jest.fn(() => stateMock);

    render(<ShowDetails />, {
      initialState: {
        show: state(),
      },
    });

    expect(screen.getByTestId("show-name")).toBeInTheDocument();
    expect(screen.getByText("The Powerpuff Girls")).toBeInTheDocument();
  });

  test("Poster should be shown", () => {
    const state = jest.fn(() => stateMock);
    const { container } = render(<ShowDetails />, {
      initialState: {
        show: state(),
      },
    });

    expect(
      container.querySelector(".show-details__poster")
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("Poster alt text should be provided", () => {
    const state = jest.fn(() => stateMock);

    const { container } = render(<ShowDetails />, {
      initialState: {
        show: state(),
      },
    });

    expect(
      container.querySelector(".show-details__poster").getAttribute("alt")
    ).toBe("Show poster");
  });
});
