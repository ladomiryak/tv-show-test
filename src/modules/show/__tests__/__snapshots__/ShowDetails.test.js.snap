// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show Details Poster should be shown 1`] = `
<div>
  <section
    class="show-details"
  >
    <img
      alt="Show poster"
      class="show-details__poster"
      src="http://static.tvmaze.com/uploads/images/original_untouched/60/151357.jpg"
    />
    <div>
      <h1
        class="show-details__title"
        data-testid="show-name"
      >
        The Powerpuff Girls
         (
        2016
        ) 
        6.2
      </h1>
      <ul
        class="show-details__genres"
      >
        <li>
          Comedy
          ,
        </li>
        <li>
          Action
          ,
        </li>
        <li>
          Science-Fiction
          
        </li>
      </ul>
      <p
        class="show-details__description"
      >
        <p>
          The city of Townsville may be a beautiful, bustling metropolis, but don't be fooled! There's evil afoot! And only three things can keep the bad guys at bay: Blossom, Bubbles and Buttercup, three super-powered little girls, known to their fans (and villains everywhere) as 
          <b>
            The Powerpuff Girls
          </b>
          . Juggling school, bedtimes, and beating up giant monsters may be daunting, but together the Powerpuff Girls are up to the task. Battling a who's who of evil, they show what it really means to "fight like a girl."
        </p>
      </p>
    </div>
  </section>
</div>
`;
